staging/src/k8s.io/apiextensions-apiserver/pkg/registry/customresource/strategy.go

```diff
@@ -63,9 +58,35 @@ func (a customResourceDefinitionStorageStrategy) NamespaceScoped() bool {
 }
 
 func (customResourceDefinitionStorageStrategy) PrepareForCreate(ctx genericapirequest.Context, obj runtime.Object) {
+	customResourceObject, _ := obj.(*unstructured.Unstructured)
+	metadata := customResourceObject.UnstructuredContent()["metadata"]
+
+	// convert metadata subtree to metadata struct and back
+	// this is done to make sure we drop unknown fields from metadata
+	// if there are errors, we don't convert and fail in ValidateCreate and ValidateUpdate
+	in := metadata.(map[string]interface{})
```

Comment:
`, ok` and short-circuit... hard casting here can panic

Reply:
probably want a helper that does this coercion so we can call it from PrepareForCreate, PrepareForUpdate, and validation functions

that would allow unit testing various scenarios as well

I had a WIP doing something similar in https://github.com/liggitt/kubernetes/commits/crd-coercion in case it's useful.

Reply:
also, based on the issues I was hitting, I don't think fromUnstructured does the nil -> "" conversion we wanted

Reply:
`structuredMetadata` for the `ValidateMetadata` test:

```
&ObjectMeta{Name:foo,GenerateName:,Namespace:not-the-default,SelfLink:/apis/mygroup.example.com/v1beta1/namespaces/not-the-default/noxus/foo,UID:169ff77e-158a-11e8-9dd2-54e1ad6c2d05,ResourceVersion:6,Generation:0,CreationTimestamp:2018-02-19 21:02:26 +0530 IST,DeletionTimestamp:<nil>,DeletionGracePeriodSeconds:nil,Labels:map[string]string{qux: ,},Annotations:map[string]string{},OwnerReferences:[],Finalizers:[],ClusterName:,Initializers:nil,}
```

Reply:
> also, based on the issues I was hitting, I don't think fromUnstructured does the nil -> "" conversion we "wanted"

Also, from this test data - https://github.com/liggitt/kubernetes/commit/51965f9e381e9588d918ad1281bf94b1bec65c58#diff-1b3f35816cc25f4fd5cd83f1c18c013fR327:

The unstructured object is:
```
map[bb:<nil> bc:<nil> bd:<nil> ba:map[aa:<nil> ab:<nil> ac:<nil>]]
```

and both `newObj` and `unmarshalledObj` are:

```
{"ba":{},"bb":"","bc":null,"bd":null}
```

which seems right to me. @liggitt thoughts?